File: cpp/Sorting/quickSort_v1.cpp
#include <iostream>
#include <vector>
using namespace std;

#define vi vector<int>

void quick_sort(vi &arr){
    int n = arr.size();
    if(n<=1) return;

    vi left,right;
    int mid = n>>1;
    int mid_elm=arr[mid];

    for(int i=0; i<n; ++i){
        if(i==mid) continue;
        arr[i]<=mid_elm?left.push_back(arr[i]):right.push_back(arr[i]);
    }
    quick_sort(left);
    quick_sort(right);

    int i=0;
    for(int x:left) arr[i++] = x;
    arr[i++] = mid_elm;
    for(int x:right) arr[i++] = x;
}

int main(){
    vi arr = {2,4,6,8,7,3,9,1,5,4,9,2,4,0,1,3,5};
    quick_sort(arr);
    for(int x: arr) cout << x << ' ';
    cout << '\n';
}

File: python/Sorting/quickSort_v1.py
import sys
sys.stdin = open("input.txt","r")

def quick_sort(arr):
    n = len(arr)
    if n<=1: return

    mid = n>>1
    mid_elm = arr[mid]
    left = list()
    right = list()
    for i in range(n):
        if i==mid: continue
        left.append(arr[i]) if(arr[i]<=mid_elm) else right.append(arr[i])
    
    quick_sort(left)
    quick_sort(right)

    i = 0
    for x in left:
        arr[i] = x
        i+=1
    arr[i] = mid_elm
    i+=1
    for x in right:
        arr[i] = x
        i+=1

arr = list(map(int,input().split()))
quick_sort(arr)
print(arr)